initrd/etc/functions

```diff
@@ -1,6 +1,390 @@
 #!/bin/bash
-# Shell functions for most initialization scripts
-. /etc/ash_functions
+
+# ------- Start of functions coming from /etc/ash_functions
+
+die() {
+	if [ "$CONFIG_DEBUG_OUTPUT" = "y" ];then
+		echo -e " !!! ERROR: $* !!!" | tee -a /tmp/debug.log /dev/kmsg > /dev/null;
+	else
+		echo -e >&2 "!!! ERROR: $* !!!";
+	fi
+	sleep 2;
+	exit 1;
+}
+
+warn() {
+	if [ "$CONFIG_DEBUG_OUTPUT" = "y" ];then
+		echo -e " *** WARNING: $* ***" | tee -a /tmp/debug.log /dev/kmsg > /dev/null;
+	else
+		echo -e >&2 " *** WARNING: $* ***";
+	fi
+	sleep 1;
+}
+
+DEBUG() {
+	if [ "$CONFIG_DEBUG_OUTPUT" = "y" ]; then
+		# fold -s -w 960 will wrap lines at 960 characters on the last space before the limit
+		echo "DEBUG: $*" | fold -s -w 960 | while read line; do
+			echo "$line" | tee -a /tmp/debug.log /dev/kmsg >/dev/null
+		done
+	fi
+}
+
+TRACE() {
+	if [ "$CONFIG_ENABLE_FUNCTION_TRACING_OUTPUT" = "y" ];then
+		echo "TRACE: $*" | tee -a /tmp/debug.log /dev/kmsg > /dev/null;
+	fi
+}
+
+# Write directly to the debug log (but not kmsg), never appears on console
+LOG() {
+	# if not CONFIG_QUIET_MODE=y, output to console. If not, output to debug.log
+	if [ "$CONFIG_DEBUG_OUTPUT" = "y" ]; then
+		DEBUG "$*"
+	elif [ "$CONFIG_QUIET_MODE" = "y" ]; then	
+		# if in quiet mode, output solely to debug.log
+		echo "$*" >> /tmp/debug.log
+	else
+		# if not in quiet mode, output to console
+		echo "$*"
```

Comment:
This needs to go to stderr, not stdout.  That's why the DO_WITH_DEBUG messed up the boot menu.  https://github.com/linuxboot/heads/pull/1875/commits/618ff26d28edd55faf498563d293842f41124c71 was a workaround for this

Reply:
Actually, it needs to go to something like /dev/console, not stdout or stderr.  This is used by DO_WITH_DEBUG which is not supposed to affect stdout or stderr.

Reply:
Actually actually, this shouldn't go to the console at all, the entire intent of LOG() was just to write to the log file for really verbose debugging information.  (That's what the comment above says, and that's why DO_WITH_DEBUG uses this.)  Why are you trying to make it go to the console at all?

Reply:
> Actually actually, this shouldn't go to the console at all, the entire intent of LOG() was just to write to the log file for really verbose debugging information.  (That's what the comment above says, and that's why DO_WITH_DEBUG uses this.)  Why are you trying to make it go to the console at all?

I wanted to use log to log output either to console(informational), file (quiet) or both (debug) per LOG calls.


Putting back to draft.

Reply:
@JonathonHall-Purism https://github.com/linuxboot/heads/pull/1875/commits/930d3e611481e4ef3583853f2367b7121ea5d264 fixes this

Reply:
@JonathonHall-Purism : op modified with log level clarifications.